spark/src/main/java/org/apache/iceberg/actions/PlanScanAction.java

```diff
@@ -0,0 +1,271 @@
+    JavaSparkContext jsc = JavaSparkContext.fromSparkContext(spark.sparkContext());
+
+    /*
+    Since we aren't filtering entries on manifest read, filter file entries using Partition information
+    In the normal path this happens with the manifest reader's entries code, filtering elements before we
+    get to this stage.
+     */
+
+    // Build cache of partition evaluators
+    ImmutableMap.Builder<Integer, Evaluator> evalMapBuilder = ImmutableMap.builder();
+    specsById.entrySet().stream().forEach(entry ->
+        evalMapBuilder.put(entry.getKey(),
+            new Evaluator(entry.getValue().partitionType(),
+                Projections.inclusive(entry.getValue(), isCaseSensitive).project(scanFilter))));
+
+    Map<Integer, Evaluator> partitionEvaluatorsById = evalMapBuilder.build();
+    Broadcast<Map<Integer, Evaluator>> broadcastPartitionEvaluators = jsc.broadcast(partitionEvaluatorsById);
+
+    Broadcast<InclusiveMetricsEvaluator> broadcastMetricsEvaluator =  jsc.broadcast(
+        new InclusiveMetricsEvaluator(scan.schema(), scanFilter, isCaseSensitive));
+
+    // Cache residual information and Partition spec information
+    final Broadcast<Map<Integer, PartitionSpec>> broadcastSpecsById = jsc.broadcast(specsById);
+
+    Types.StructType partitionStruct = DataFile.getType(scan.table().spec().partitionType());
+    StructType dataFileSchema = (StructType) dataFileEntries.schema().apply("data_file").dataType();
+
+    // Evaluate all files based on their partition info and collect the rows back locally
+    Dataset<Row> scanTaskDataset = dataFileEntries.mapPartitions(
+        (MapPartitionsFunction<Row, Row>) it -> {
+          SparkDataFile container = new SparkDataFile(partitionStruct, dataFileSchema);
+          return Streams.stream(it)
+              .filter(row -> {
+                Row dataFile = row.getAs("data_file");
+                SparkDataFile file = container.wrap(dataFile);
+                return broadcastPartitionEvaluators.getValue().get(file.specId()).eval(file.partition()) &&
+                    broadcastMetricsEvaluator.getValue().eval(file);
+              }).iterator();
+        }, RowEncoder.apply(dataFileEntries.schema()));
+
+    LoadingCache<Integer, SpecCacheEntry> specCache =
+        Caffeine.newBuilder().build((CacheLoader<Integer, SpecCacheEntry> & Serializable) specId -> {
+          PartitionSpec spec = broadcastSpecsById.getValue().get(specId);
+          Expression filter = ignoreResiduals ? Expressions.alwaysTrue() : scanFilter;
+          return new SpecCacheEntry(SchemaParser.toJson(spec.schema()), PartitionSpecParser.toJson(spec),
+              ResidualEvaluator.of(spec, filter, isCaseSensitive));
+        });
+
+    // Build delete index locally
+    DeleteFileIndex deleteFileIndex = buildDeleteFileIndex();
+    SparkDataFile container = new SparkDataFile(partitionStruct, dataFileSchema);
+    List<FileScanTask> tasks = scanTaskDataset.collectAsList().stream().map(row -> {
```

Comment:
We aren't building tasks remotely because DeleteFileIndex and DeleteFileBuilder are both setup in ways that make serialization difficult. DeleteFileIndex uses thread locals and deletefilebuilder owns a ExecutionService. We can probably parallelize one of these in the future though and move all task construction remotely as well. I'm not sure there is a huge benefit to this though.

Reply:
scnaTaskDataset is always going to be relatively small in return value right?

Reply:
It will be no larger than one we would have to make locally, so yes there is a possibility for an OOM here, but no more so than during the normal planning path.